drivers/display/display_st7735r.c

```diff
@@ -0,0 +1,494 @@
+}
+
+static int st7735r_write(const struct device *dev,
+			 const uint16_t x,
+			 const uint16_t y,
+			 const struct display_buffer_descriptor *desc,
+			 const void *buf)
+{
+	struct st7735r_data *data = (struct st7735r_data *)dev->data;
+	const uint8_t *write_data_start = (uint8_t *) buf;
+	struct spi_buf tx_buf;
+	struct spi_buf_set tx_bufs;
+	uint16_t write_cnt;
+	uint16_t nbr_of_writes;
+	uint16_t write_h;
+
+	__ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
+	__ASSERT((desc->pitch * ST7735R_PIXEL_SIZE * desc->height)
+		 <= desc->buf_size, "Input buffer too small");
+
+	LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)",
+		desc->width, desc->height, x, y);
+	st7735r_set_mem_area(data, x, y, desc->width, desc->height);
+
+	if (desc->pitch > desc->width) {
+		write_h = 1U;
+		nbr_of_writes = desc->height;
+	} else {
+		write_h = desc->height;
+		nbr_of_writes = 1U;
+	}
+
+	st7735r_transmit(data, ST7735R_CMD_RAMWR,
+			 (void *) write_data_start,
+			 desc->width * ST7735R_PIXEL_SIZE * write_h);
+
+	tx_bufs.buffers = &tx_buf;
+	tx_bufs.count = 1;
+
+	write_data_start += (desc->pitch * ST7735R_PIXEL_SIZE);
+	for (write_cnt = 1U; write_cnt < nbr_of_writes; ++write_cnt) {
+		tx_buf.buf = (void *)write_data_start;
+		tx_buf.len = desc->width * ST7735R_PIXEL_SIZE * write_h;
+		spi_write(data->spi_dev, &data->spi_config, &tx_bufs);
+		write_data_start += (desc->pitch * ST7735R_PIXEL_SIZE);
+	}
+
+	return 0;
+}
+
+static void *st7735r_get_framebuffer(const struct device *dev)
+{
+	return NULL;
+}
+
+static int st7735r_set_brightness(const struct device *dev,
+				  const uint8_t brightness)
+{
+	return -ENOTSUP;
+}
+
+static int st7735r_set_contrast(const struct device *dev,
+				const uint8_t contrast)
+{
+	return -ENOTSUP;
+}
+
+static void st7735r_get_capabilities(const struct device *dev,
+				     struct display_capabilities *capabilities)
+{
+	struct st7735r_config *config = (struct st7735r_config *)dev->config;
+
+	memset(capabilities, 0, sizeof(struct display_capabilities));
+	capabilities->x_resolution = config->width;
+	capabilities->y_resolution = config->height;
+	if (config->mdac & ST7735R_MADCTL_BGR) {
+		capabilities->supported_pixel_formats = PIXEL_FORMAT_BGR_565;
+		capabilities->current_pixel_format = PIXEL_FORMAT_BGR_565;
+	} else {
+		capabilities->supported_pixel_formats = PIXEL_FORMAT_RGB_565;
+		capabilities->current_pixel_format = PIXEL_FORMAT_RGB_565;
+	}
+	capabilities->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+static int st7735r_set_pixel_format(const struct device *dev,
+				    const enum display_pixel_format pixel_format)
+{
+	struct st7735r_config *config = (struct st7735r_config *)dev->config;
+
+	if ((pixel_format == PIXEL_FORMAT_RGB_565) &&
+	    (~config->mdac & ST7735R_MADCTL_BGR)) {
+		return 0;
+	}
+	if ((pixel_format == PIXEL_FORMAT_BGR_565) &&
+	    (config->mdac & ST7735R_MADCTL_BGR)) {
+		return 0;
+	}
+
+	LOG_ERR("Pixel format change not implemented");
+	return -ENOTSUP;
```

Comment:
I'm not sure if this function should just return `-ENOTSUP` in anycase, as the operation is not supported. @vanwinkeljan ?

Reply:
It is done the same way in other drivers, in case a supported pixel format is set `zero` is returned else `ENOTSUP`

Reply:
If it does always support -ENOTSUP there might be application code which won't work. At least with this approach the supported settings can be set - even though it doesn't make much sense in the driver.